File: sdk/js/gladys-sensor-sdk/test/CommandDispatcher.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { CommandDispatcher } from "../src/CommandDispatcher";
import { ComponentState } from "../src/generated/common";
import { Command } from "../src/generated/orchestrator";

describe("CommandDispatcher", () => {
  let dispatcher: CommandDispatcher;

  beforeEach(() => {
    dispatcher = new CommandDispatcher();
  });

  it("dispatches START and sets ACTIVE", async () => {
    dispatcher.onStart(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_START, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
    expect(result.errorMessage).toBeUndefined();
  });

  it("dispatches STOP and sets STOPPED", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher.onStop(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_STOP, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_STOPPED);
  });

  it("dispatches PAUSE and sets PAUSED", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher.onPause(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_PAUSE);

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_PAUSED);
  });

  it("dispatches RESUME and sets ACTIVE", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_PAUSED);
    dispatcher.onResume(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_RESUME);

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches HEALTH_CHECK without state change", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher.onHealthCheck(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_HEALTH_CHECK, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
    expect(dispatcher.getState()).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches RECOVER and sets ACTIVE", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ERROR);
    dispatcher.onRecover(async () => null);

    const result = await dispatcher.dispatch(Command.COMMAND_RECOVER, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("handler override state is respected", async () => {
    dispatcher.onStart(async () => {
      return ComponentState.COMPONENT_STATE_STARTING;
    });

    const result = await dispatcher.dispatch(Command.COMMAND_START, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_STARTING);
  });

  it("handler exception sets ERROR", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher.onStop(async () => {
      throw new Error("shutdown failed");
    });

    const result = await dispatcher.dispatch(Command.COMMAND_STOP, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ERROR);
    expect(result.errorMessage).toBe("shutdown failed");
  });

  it("HEALTH_CHECK exception preserves state", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher.onHealthCheck(async () => {
      throw new Error("check failed");
    });

    const result = await dispatcher.dispatch(Command.COMMAND_HEALTH_CHECK, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
    expect(result.errorMessage).toBe("check failed");
  });

  it("unregistered command keeps state and returns error message", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_PAUSED);
    const result = await dispatcher.dispatch(Command.COMMAND_RELOAD);

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_PAUSED);
    expect(result.errorMessage).toBe(
      `No handler registered for ${Command.COMMAND_RELOAD}`
    );
  });

  it("error handler is called on exception", async () => {
    let errorHandlerCalled = false;
    let capturedCommand: Command | undefined;
    let capturedError: Error | undefined;
    let capturedState: ComponentState | undefined;

    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher
      .onStop(async () => {
        throw new Error("stop broke");
      })
      .onCommandError((cmd, err, state) => {
        errorHandlerCalled = true;
        capturedCommand = cmd;
        capturedError = err;
        capturedState = state;
        return null; // Accept ERROR
      });

    const result = await dispatcher.dispatch(Command.COMMAND_STOP, {});

    expect(errorHandlerCalled).toBe(true);
    expect(capturedCommand).toBe(Command.COMMAND_STOP);
    expect(capturedError?.message).toBe("stop broke");
    expect(capturedState).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ERROR);
  });

  it("error handler can override ERROR state", async () => {
    dispatcher.setState(ComponentState.COMPONENT_STATE_ACTIVE);
    dispatcher
      .onStop(async () => {
        throw new Error("stop broke");
      })
      .onCommandError((_cmd, _err, currentState) => {
        return currentState; // Stay in current state instead of ERROR
      });

    const result = await dispatcher.dispatch(Command.COMMAND_STOP, {});

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });
});

File: sdk/java/gladys-sensor-sdk/src/test/java/com/gladys/sensor/CommandDispatcherTest.java
package com.gladys.sensor;

import com.google.protobuf.Struct;
import gladys.v1.Common;
import gladys.v1.Orchestrator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

class CommandDispatcherTest {

    private AtomicBoolean errorHandlerCalled;

    @BeforeEach
    void setUp() {
        errorHandlerCalled = new AtomicBoolean(false);
    }

    @Test
    void testDispatchStartSetsActive() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchStopSetsStopped() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStop(args -> null)
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_STOP, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchPauseSetsPaused() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onPause(() -> null)
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_PAUSE, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_PAUSED, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_PAUSED, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchResumeFromPausedSetsActive() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onPause(() -> null)
                .onResume(() -> null)
                .build();

        // First pause
        dispatcher.dispatch(Orchestrator.Command.COMMAND_PAUSE, Struct.getDefaultInstance());
        assertEquals(Common.ComponentState.COMPONENT_STATE_PAUSED, dispatcher.getCurrentState());

        // Then resume
        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_RESUME, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchHealthCheckUnchanged() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .onHealthCheck(args -> null)
                .build();

        // Set state to ACTIVE first
        dispatcher.dispatch(Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());

        // Health check should not change state
        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_HEALTH_CHECK, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchRecoverSetsActive() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onRecover(args -> null)
                .build();

        // Set state to ERROR first
        dispatcher.setCurrentState(Common.ComponentState.COMPONENT_STATE_ERROR);

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_RECOVER, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testHandlerOverrideState() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> Common.ComponentState.COMPONENT_STATE_STOPPED)
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, dispatcher.getCurrentState());
    }

    @Test
    void testHandlerExceptionSetsError() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> { throw new RuntimeException("startup failed"); })
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        assertTrue(result.hasError());
        assertEquals("startup failed", result.errorMessage);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ERROR, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ERROR, dispatcher.getCurrentState());
    }

    @Test
    void testHealthCheckExceptionKeepsState() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .onHealthCheck(args -> { throw new RuntimeException("check failed"); })
                .build();

        // Start first to set ACTIVE state
        dispatcher.dispatch(Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());

        // Health check failure should preserve ACTIVE state
        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_HEALTH_CHECK, Struct.getDefaultInstance());

        assertTrue(result.hasError());
        assertEquals("check failed", result.errorMessage);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testNoHandlerKeepsStateAndReturnsError() {
        CommandDispatcher dispatcher = CommandDispatcher.builder().build();

        dispatcher.setCurrentState(Common.ComponentState.COMPONENT_STATE_ACTIVE);

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        assertTrue(result.hasError());
        assertEquals("No handler registered for COMMAND_START", result.errorMessage);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals("No handler registered for COMMAND_START", dispatcher.getLastErrorMessage());
    }

    @Test
    void testErrorHandlerCalled() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> { throw new RuntimeException("boom"); })
                .onCommandError((cmd, ex, state) -> {
                    errorHandlerCalled.set(true);
                    assertEquals(Orchestrator.Command.COMMAND_START, cmd);
                    assertEquals("boom", ex.getMessage());
                    // Return STOPPED instead of default ERROR
                    return Common.ComponentState.COMPONENT_STATE_STOPPED;
                })
                .build();

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        assertTrue(errorHandlerCalled.get());
        assertTrue(result.hasError());
        assertEquals("boom", result.errorMessage);
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, dispatcher.getCurrentState());
    }

    @Test
    void testHealthCheckExceptionCallsErrorHandler() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .onHealthCheck(args -> { throw new RuntimeException("health boom"); })
                .onCommandError((cmd, ex, state) -> {
                    errorHandlerCalled.set(true);
                    assertEquals(Orchestrator.Command.COMMAND_HEALTH_CHECK, cmd);
                    assertEquals("health boom", ex.getMessage());
                    return Common.ComponentState.COMPONENT_STATE_STOPPED;
                })
                .build();

        dispatcher.dispatch(Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance());

        CommandDispatcher.DispatchResult result = dispatcher.dispatch(
                Orchestrator.Command.COMMAND_HEALTH_CHECK, Struct.getDefaultInstance());

        assertTrue(errorHandlerCalled.get());
        assertTrue(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, dispatcher.getCurrentState());
    }

    @Test
    void testErrorHandlerThrowingDoesNotEscapeDispatch() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> { throw new RuntimeException("handler failed"); })
                .onCommandError((cmd, ex, state) -> {
                    throw new RuntimeException("error handler failed");
                })
                .build();

        CommandDispatcher.DispatchResult result = assertDoesNotThrow(() ->
                dispatcher.dispatch(Orchestrator.Command.COMMAND_START, Struct.getDefaultInstance()));

        assertTrue(result.hasError());
        assertEquals("handler failed", result.errorMessage);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ERROR, result.state);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ERROR, dispatcher.getCurrentState());
    }

    @Test
    void testDispatchIsSynchronized() throws Exception {
        Method dispatchMethod = CommandDispatcher.class.getDeclaredMethod(
                "dispatch", Orchestrator.Command.class, Struct.class);
        assertTrue(Modifier.isSynchronized(dispatchMethod.getModifiers()));

        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .onStop(args -> null)
                .build();

        int threads = 8;
        int iterationsPerThread = 200;
        ExecutorService executor = Executors.newFixedThreadPool(threads);
        CountDownLatch ready = new CountDownLatch(threads);
        CountDownLatch start = new CountDownLatch(1);
        AtomicReference<Throwable> firstFailure = new AtomicReference<>();
        AtomicInteger failureCount = new AtomicInteger();

        try {
            for (int i = 0; i < threads; i++) {
                final boolean useStart = i % 2 == 0;
                executor.submit(() -> {
                    ready.countDown();
                    try {
                        start.await();
                        for (int j = 0; j < iterationsPerThread; j++) {
                            dispatcher.dispatch(
                                    useStart ? Orchestrator.Command.COMMAND_START : Orchestrator.Command.COMMAND_STOP,
                                    Struct.getDefaultInstance());
                        }
                    } catch (Throwable t) {
                        failureCount.incrementAndGet();
                        firstFailure.compareAndSet(null, t);
                    }
                });
            }

            assertTrue(ready.await(5, TimeUnit.SECONDS));
            start.countDown();
        } finally {
            executor.shutdown();
            assertTrue(executor.awaitTermination(10, TimeUnit.SECONDS));
        }

        if (firstFailure.get() != null) {
            fail("dispatch() failed during concurrent calls: " + firstFailure.get().getClass().getSimpleName());
        }
        assertEquals(0, failureCount.get(), "Expected no exceptions during concurrent dispatch calls");
    }
}
